### DISCUSSION TITLE
Magic Methods for PyPDF2

### DISCUSSION THREAD

**Discussion by MartinThoma** [py-pdf/pypdf#751] [Ideas]
I would like to use magic methods for PyPDF2 and hear your ideas about it:

## #1: `PdfFileReader.__length__` :x: 

Usage example:

```pycon
>>> from PyPDF2 import PdfFileReader
>>> reader = PdfFileReader("path/to/my.pdf")
>>> len(reader)   # the number of pages
42
```

## #2: `PageRange.__add__` :heavy_check_mark: -  #948 

Usage example:

```pycon
>>> from PyPDF2 import PageRange
>>> PageRange("1:3") + PageRange("2:5")
PageRange('1:5')
```

## #3: `PdfFileReader.__rshift__` :x: 

Put all contents of one pdf in a reader:

```python
from PyPDF2 import PdfFileReader, PdfFileWriter
reader = PdfFileReader("path/to/my.pdf")
writer = PdfFileWriter()

# New
reader >> writer

# Equivalent to the old:
for page in reader.pages:
    writer.addPage(page)
```

## #4: `PdfFileReader.__getitem__` :x: 

```python
from PyPDF2 import PdfFileReader
reader = PdfFileReader("path/to/my.pdf")

# New
page = reader[42]

# Equivalent to the old:
page = reader.pages[42]
```

## Thoughts?

Those are the ones that pop to my mind, but maybe the heavy PyPDF2 users have other operations in mind that you would like to do? What do you think about those?

**Comment by MartinThoma** [comment]
This is what I imagine: https://github.com/py-pdf/PyPDF2/pull/759 - there is still a lot of functions / attributes I would need to rename.

**Comment by sekrause** [comment]
* `PageRange.__add__` looks nice.
* `PdfFileReader.__rshift__` looks very un-pythonic and confusing. I would argue against something like that.
* I would also argue against `PdfFileReader.__length__` and `PdfFileReader.__getitem__`. I think it's better to only allow those operations on `reader.pages`. Having two different ways of doing that is against the [Zen of Python](https://peps.python.org/pep-0020/#the-zen-of-python): _"There should be one-- and preferably only one --obvious way to do it."_

**Reply by MasterOdin** [reply]
I agree with the point about `PdfFileReader.__length__` and `PdfFileReader.__getitem__` as I'd expect those (or any magic method) on `PdfFileReader` to operate directly onto the reader, and not an underlying property of it (`pages`).

**Comment by exiledkingcc** [comment]
I prefer adding a `PdfPageView` class to present all pages in pdf, then add `__length__`, `__getitem__`, `__iter__`, `__next__` for it.
